presto-prometheus/src/main/java/io/prestosql/plugin/prometheus/PrometheusMetadata.java

```diff
@@ -0,0 +1,179 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.plugin.prometheus;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import io.prestosql.spi.connector.ColumnHandle;
+import io.prestosql.spi.connector.ColumnMetadata;
+import io.prestosql.spi.connector.ConnectorMetadata;
+import io.prestosql.spi.connector.ConnectorSession;
+import io.prestosql.spi.connector.ConnectorTableHandle;
+import io.prestosql.spi.connector.ConnectorTableMetadata;
+import io.prestosql.spi.connector.ConnectorTableProperties;
+import io.prestosql.spi.connector.Constraint;
+import io.prestosql.spi.connector.ConstraintApplicationResult;
+import io.prestosql.spi.connector.SchemaTableName;
+import io.prestosql.spi.connector.SchemaTablePrefix;
+import io.prestosql.spi.connector.TableNotFoundException;
+
+import javax.inject.Inject;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+import static java.util.Objects.requireNonNull;
+
+public class PrometheusMetadata
+        implements ConnectorMetadata
+{
+    private PrometheusClient prometheusClient;
+
+    @Inject
+    public PrometheusMetadata(PrometheusClient pc)
+    {
+        prometheusClient = pc;
+        requireNonNull(prometheusClient, "client is null");
+    }
+
+    @Override
+    public List<String> listSchemaNames(ConnectorSession session)
+    {
+        return listSchemaNames();
+    }
+
+    public List<String> listSchemaNames()
+    {
+        return ImmutableList.copyOf(prometheusClient.getSchemaNames());
+    }
+
+    @Override
+    public PrometheusTableHandle getTableHandle(ConnectorSession session, SchemaTableName tableName)
+    {
+        if (!listSchemaNames(session).contains(tableName.getSchemaName())) {
+            return null;
+        }
+
+        if (prometheusClient.getTable(tableName.getSchemaName(), tableName.getTableName()) == null) {
+            return null;
+        }
+
+        return new PrometheusTableHandle(tableName.getSchemaName(), tableName.getTableName());
+    }
+
+    @Override
+    public ConnectorTableMetadata getTableMetadata(ConnectorSession session, ConnectorTableHandle table)
+    {
+        return getTableMetadata(((PrometheusTableHandle) table).toSchemaTableName());
+    }
+
+    @Override
+    public List<SchemaTableName> listTables(ConnectorSession session, Optional<String> optionalSchemaName)
+    {
+        Set<String> schemaNames = optionalSchemaName.map(ImmutableSet::of)
+                .orElseGet(() -> ImmutableSet.copyOf(prometheusClient.getSchemaNames()));
+
+        ImmutableList.Builder<SchemaTableName> builder = ImmutableList.builder();
+        for (String schemaName : schemaNames) {
+            for (String tableName : prometheusClient.getTableNames(schemaName)) {
```

Comment:
You can use `schemaNames.stream().flatMap(...).collect(toImmutableList())`

Reply:
Done in c46796a2